brainiak/isfc.py

```diff
@@ -1,226 +1,1104 @@
+    
+    if group_assignment and len(group_assignment) != n_subjects:
+        raise ValueError(f"Group assignments ({len(group_assignment)}) "
+                         f"do not match number of subjects ({n_subjects})!")
+    
+    # Set up group selectors for two-group scenario
+    if group_assignment and len(np.unique(group_assignment)) == 2:
+        n_groups = 2
+        
+        # Get group labels and counts
+        group_labels = np.unique(group_assignment)
+        groups = {group_labels[0]: group_assignment.count(group_labels[0]),
+                  group_labels[1]: group_assignment.count(group_labels[1])}
+
+        # For two-sample pairwise approach set up selector from matrix
+        if pairwise == True:
+            # Sort the group_assignment variable if it came in shuffled
+            # so it's easier to build group assignment matrix
+            sorter = np.array(group_assignment).argsort()
+            unsorter = np.array(group_assignment).argsort().argsort()
+            
+            # Populate a matrix with group assignments                                 
+            group_matrix = np.vstack((np.hstack((np.full((groups[group_labels[0]],
+                                                          groups[group_labels[0]]),
+                                                         group_labels[0]),
+                                                 np.full((groups[group_labels[0]],
+                                                          groups[group_labels[1]]),
+                                                        np.nan))),
+                                      np.hstack((np.full((groups[group_labels[1]],
+                                                          groups[group_labels[0]]),
+                                                         np.nan),
+                                                 np.full((groups[group_labels[1]],
+                                                          groups[group_labels[1]]),
+                                                         group_labels[1])))))
+            np.fill_diagonal(group_matrix, np.nan)
+            
+            # Unsort matrix and squareform to create selector
+            group_selector = squareform(group_matrix[unsorter, :][:, unsorter],
+                                        checks=False)
+            
+        # If leave-one-out approach, just user group assignment as selector
+        elif pairwise == False:
+            group_selector = group_assignment
+    
+    # Manage one-sample and incorrect group assignments
+    elif not group_assignment or len(np.unique(group_assignment)) == 1:
+        n_groups = 1
+        
+        # If pairwise initialize matrix of ones for sign-flipping
+        ones_matrix = np.ones((n_subjects, n_subjects))
+        
+    elif len(np.unique(group_assignment)) > 2:
+        raise ValueError("This test is not valid for more than "
+                         f"2 groups! (got {n_groups})")
```

Comment:
`n_groups` is referred before assignment here. Consider using `len(np.unique(group_assignment))` instead?